state/execution.go

```diff
@@ -254,6 +255,54 @@ func (blockExec *BlockExecutor) Commit(
 	return res.Data, res.RetainHeight, err
 }
 
+func (blockExec *BlockExecutor) CheckBlock(block *types.Block) error {
+	commitInfo, byzVals := getBlockValidatorInfo(
+		block.Time,
+		block.Height,
+		block.LastCommit,
+		block.Evidence.Evidence,
+		blockExec.db,
+	)
+	pbh := block.Header.ToProto()
+	txs := make([][]byte, 0, len(block.Txs))
+	for _, tx := range block.Txs {
+		txs = append(txs, tx)
+	}
+
+	resp, err := blockExec.proxyApp.CheckBlockSync(abcix.RequestCheckBlock{
+		Height:              block.Height,
+		Hash:                block.Hash(),
+		Header:              *pbh,
+		LastCommitInfo:      commitInfo,
+		ByzantineValidators: byzVals,
+		Txs:                 txs,
+	})
+
+	if err != nil {
+		return err
+	}
+	if resp.Code != 0 {
+		return fmt.Errorf("application error during CheckBlock, code: %d", resp.Code)
+	}
+	for _, tx := range resp.DeliverTxs {
+		if tx.Code != 0 {
+			return fmt.Errorf("invalid transaction, code: %d", tx.Code)
+		}
+	}
+	if !bytes.Equal(resp.ResultHash, block.Header.LastResultsHash.Bytes()) {
+		return fmt.Errorf(
+			"resultHash mismatch\nResultHash in ResponseCheckBlock: %X\n ResultHash in block header: %X",
```

Comment:
error message should simple and generic. you can error log the details while returning a more concise error message.